src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+                    // Skip empty file. The checksum of this cf file should be 0 and
+                    // this is checked when loading the snapshot meta.
+                    continue;
+                }
+                try!(check_file_size_and_checksum(&cf_file.path, cf_file.size, cf_file.checksum));
+            }
+            Ok(())
+        }
+
+        fn switch_to_cf_file(&mut self, cf: &str) -> io::Result<()> {
+            match self.cf_files.iter().position(|x| x.cf == cf) {
+                Some(index) => {
+                    self.cf_index = index;
+                    Ok(())
+                }
+                None => Err(io::Error::new(ErrorKind::Other, format!("fail to find cf {}", cf))),
+            }
+        }
+
+        fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+            let mut cf_file = &mut self.cf_files[self.cf_index];
+            let mut writer = cf_file.sst_writer.as_mut().unwrap();
+            if let Err(e) = writer.add(k, v) {
+                return Err(io::Error::new(ErrorKind::Other, e));
+            }
+            cf_file.kv_count += 1;
+            Ok(())
+        }
+
+        fn save_cf_files(&mut self) -> io::Result<()> {
+            for cf_file in &mut self.cf_files {
+                if plain_file_used(cf_file.cf) {
+                    let _ = cf_file.file.take();
+                } else if cf_file.kv_count == 0 {
+                    let _ = cf_file.sst_writer.take().unwrap();
+                } else {
+                    let mut writer = cf_file.sst_writer.take().unwrap();
+                    if let Err(e) = writer.finish() {
+                        return Err(io::Error::new(ErrorKind::Other, e));
+                    }
+                }
+                let size = try!(get_file_size(&cf_file.tmp_path));
+                if size > 0 {
+                    try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                    cf_file.size = size;
+                    cf_file.checksum = try!(calc_crc32(&cf_file.path));
+                } else {
+                    // Clean up the `tmp_path` if this cf file is empty.
+                    delete_file_if_exist(&cf_file.tmp_path);
+                }
+            }
+            Ok(())
+        }
+
+        fn save_meta_file(&mut self) -> RaftStoreResult<()> {
+            let mut v = vec![];
+            box_try!(self.meta_file.meta.write_to_vec(&mut v));
+            {
+                let mut f = self.meta_file.file.take().unwrap();
+                try!(f.write_all(&v[..]));
+                try!(f.flush());
+            }
+            try!(fs::rename(&self.meta_file.tmp_path, &self.meta_file.path));
+            Ok(())
+        }
+
+        fn do_build(&mut self,
+                    snap: &DbSnapshot,
+                    region: &Region,
+                    stat: &mut SnapshotStatistics)
+                    -> RaftStoreResult<()> {
+            if self.exists() {
+                match self.validate() {
```

Comment:
What I mean is the file only needs to be checked once. Either checks can be removed.